language/move-vm/runtime/src/interpreter.rs

```diff
@@ -869,6 +1033,504 @@ impl Frame {
             })
     }
 
+    fn pre_execution_type_stack_transition(
+        local_tys: &[Type],
+        locals: &Locals,
+        _ty_args: &[Type],
+        resolver: &Resolver,
+        interpreter: &mut Interpreter,
+        instruction: &Bytecode,
+    ) -> PartialVMResult<()> {
+        match instruction {
+            // Call instruction will be checked at execute_main.
+            Bytecode::Call(_) | Bytecode::CallGeneric(_) => (),
+            Bytecode::BrFalse(_) | Bytecode::BrTrue(_) => {
+                interpreter.operand_stack.pop_ty()?;
+            }
+            Bytecode::Branch(_) => (),
+            Bytecode::Ret => {
+                for (idx, ty) in local_tys.iter().enumerate() {
+                    if !locals.is_invalid(idx)? {
+                        check_ability(resolver.loader().abilities(&ty)?.has_drop())?;
+                    }
+                }
+            }
+            Bytecode::Abort => {
+                interpreter.operand_stack.pop_ty()?;
+            }
+            // StLoc needs to check before execution as we need to check the drop ability of values.
+            Bytecode::StLoc(idx) => {
```

Comment:
I see, I did not realize there were two sets of checks. What would go wrong if we did all the checks either before execution or after (minus the gas issue mentioned on "before", which seems fairly minor)

Reply:
For some of the operations, the cost of resolving the types seems non trivial to me. That's the main reason that makes me feel that we should only do the checks after we've charged proper gas here. Would be great if @vgao1996 can weigh in your gas experience here. To answer your question:

- Calling everything before will work, but just that we might run the checks before we charge gas, and some of those checks could be more expensive. 
- Calling everything after won't work for all control flow operators and `StLoc`. The reason is that for control flow oeprators the after path won't be invoked at all. The `StLoc` is a very interesting case here. The reason why it won't work is because we need to check if the local we can potentially overwrite has the drop ability, and the only way to check that is before we execute the instruction.

Reply:
Thanks for the explanation. My bias would be for clarity/readability--these checks should only be run without charging gas in the (hopefully rare/impossible) event that that there's a bug in the bytecode verifier, and the cost of any individual instruction is (hopefully) very small. So I think the scenario that splitting the checks is protecting against is both unlikely and low consequence. Whereas a bug in the checks that happens due to the before/after split would be unlikely and high consequence.

This is subjective and I'm ok with whatever you decide on this.